### PR TITLE
add ripple, disabled.  use renderer.

### PR DESCRIPTION
**Pull request by jefrog1844**
Please review

### CODE REVIEW

File: projects/mui-angular/button/button.component.ts [hunk 1]
```diff
@@ -1,36 +1,82 @@
...
-
+import { Component, ElementRef, Input, ViewChild, HostListener, Renderer2 } from '@angular/core';
 
 @Component({
   selector: 'mui-button',
   template: `
-    <button class="mui-btn" #button>
+    <button class="mui-btn" #button [disabled]="disabled">
       <ng-content></ng-content>
+      <span class="mui-btn__ripple-container" #container><span class="mui-ripple" #ripple></span></span>
```
Comment (amorey) [inline projects/mui-angular/button/button.component.ts:8]: Is the `#container` reference being used?

File: projects/mui-angular/button/button.component.ts [hunk 2]
```diff
@@ -1,36 +1,82 @@
...
   @Input() variant: string = '';
   @Input() color: string = '';
   @Input() size: string = '';
-  
-  constructor() {
+  @Input() disabled?: boolean = false;
+
+  @HostListener('mousedown', ['$event', '$event.currentTarget'])
+  mouseDown(event: MouseEvent) {
+    if(this.disabled) return;
```
Comment (amorey) [inline projects/mui-angular/button/button.component.ts:24]:
```suggestion
    if (this.disabled) return;
```

File: projects/mui-angular/button/button.component.ts [hunk 3]
```diff
@@ -1,36 +1,82 @@
...
+        this.renderer.addClass(this.button.nativeElement, 'mui-btn--' + attrVal);
       }
     });
   }
+
+  private getFadeOutAnimation(event) {
+    var offsetWidth, offsetHeight, offsetTop, offsetLeft, radius, diameter, width, height, left, top;
+
+    // get (x, y) position of click
+    offsetWidth = this.button.nativeElement.offsetWidth;
```
Comment (amorey) [inline projects/mui-angular/button/button.component.ts:51]: You can save some keystrokes by saving references to `this.button.nativeElement` and `this.ripple.nativeElement`

File: projects/mui-angular/button/button.component.ts [hunk 4]
```diff
@@ -1,36 +1,82 @@
...
+  constructor(private renderer: Renderer2) {
   }
 
   ngAfterViewInit() {
     // set button styles
     ['variant', 'color', 'size'].forEach(attrName => {
       let attrVal = this[attrName];
       if (attrVal) {
-        this.button.nativeElement.classList.add('mui-btn--' + attrVal);
+        this.renderer.addClass(this.button.nativeElement, 'mui-btn--' + attrVal);
```
Comment (amorey) [inline projects/mui-angular/button/button.component.ts:42]: 👍 Woot! The renderer makes class manipulation much easier.

**Comment by amorey** [comment]
Looks great! I added some suggestions in-line. Two other comments:

- If possible it would be nice to add some tests as well. I've found that even simple tests can help you to improve your code and make it more maintainable in the long-run.
- Are you familiar with [Airbnb's JavaScript styleguide](https://github.com/airbnb/javascript)? I use it for my javascript code and have found it to be very useful. It has some tips that would be helpful here too.

File: projects/mui-angular/button/button.component.ts [hunk 5]
```diff
@@ -1,36 +1,82 @@
...
 
-
-export class ButtonComponent implements OnInit {
+export class ButtonComponent {
   @ViewChild('button') button: ElementRef;
+  @ViewChild('ripple') ripple: ElementRef;
 
   @Input() variant: string = '';
   @Input() color: string = '';
   @Input() size: string = '';
```
Comment (amorey) [inline projects/mui-angular/button/button.component.ts:19]: Should we make these inputs `optional?` as well?

**Comment by jefrog1844** [comment]
Got your feedback and I will submit an update based on your feedback.

1. Is there an editor you use that has the airbnb styleguide incorporated?
2. Are you fine with the [disabled]="true"?  To use "disabled" instead of
"[disabled]=true], I would have to use a set method on the input with
another private const to bind the disabled tag to?

Thanks

Jeff

File: projects/mui-angular/button/button.component.ts (hunk 1 above)
Comment (jefrog1844) [reply projects/mui-angular/button/button.component.ts:8]: No, will remove.

File: projects/mui-angular/button/button.component.ts (hunk 2 above)
Comment (jefrog1844) [reply projects/mui-angular/button/button.component.ts:24]: Done

File: projects/mui-angular/button/button.component.ts (hunk 3 above)
Comment (jefrog1844) [reply projects/mui-angular/button/button.component.ts:51]: added private variables and set in ngAfterViewInit

File: projects/mui-angular/button/button.component.ts (hunk 4 above)
Comment (jefrog1844) [reply projects/mui-angular/button/button.component.ts:42]: even nicer with cached elements!

File: projects/mui-angular/button/button.component.ts (hunk 5 above)
Comment (jefrog1844) [reply projects/mui-angular/button/button.component.ts:19]: Yes.

**Comment by amorey** [comment]
For help with the styleguide, it looks like you can use eslint to do Airbnb styleguide linting from the command line:
https://www.npmjs.com/package/eslint-config-airbnb

You can also integrate the linter with some popular IDE's:
- Atom (https://www.acuriousanimal.com/blog/2016/08/14/configuring-atom-with-eslint)
- SublimeText (https://www.cuga-moylan.com/blog/setting-up-eslint-with-airbnb-styleguide-in-sublime-text-3/)

For the pull request, if you push your changes to your github repo then it should update this PR automatically.

**Comment by jefrog1844** [comment]
I have the changes done and have written the test class, but I’m not able to get the test classes to run except those in the src directory that are the generic component and service you created.  Any ideas?

Thanks

Jeff

**Comment by jefrog1844** [comment]
I applied the changes you requested and got the unit test to work.  I'm
going to see about getting eslint working now.

Jeff

**Comment by amorey** [comment]
Awesome! Looks great! Let me know when you get the linter working.

**Comment by jefrog1844** [comment]
the air bnb suggested additional formatting so I applied those.  Also, found a cleaner way to check class list.  I found that the disabled check appeared to do nothing so I removed it for now.

**Comment by jefrog1844** [comment]
I’m struggling on linter, it just looks too messy as everyone has a different way to include it in their code editor.  I had it working in vs code, but now it is not and it altered angular.json, tsconfig.json, and it conflicted with tslint.json, on and on.

I’m not sure if you want to include lint of dependencies in the library or not.  If you have an easy suggest, I’m open to it.

Also, on GitHub: I made another submit.  From what I understand, I have a forked copy that I can keep updated from your master.  From my fork, I can create a branch like I did for button.  The way I understand it now is that I should wait until you merge the button branch and then fetch from upstream to update my fork.  Then I can create another branch for the next feature.  Correct?

Thanks for the help.

Jeff

**Comment by amorey** [comment]
Same here. I tried setting up eslint and ran into problems. For now, maybe we should stick with Angular's default linter. Just run `ng lint` and it will make some suggestions.

Your github workflow sounds good (1. create feature branch 2. submit pull request 3. merge upstream changes into master 4. repeat).

**Comment by jefrog1844** [comment]
Sounds good.  I'll test ng lint and I've got a good idea of what you're
looking for now.

What component do you want next?  You'll need to merge the button branch in
first correct?

Thanks

Jeff

**Comment by amorey** [comment]
I think you can probably submit another PR independent of the button merge but we might as well merge first. How about knocking off some easy ones next, like container, divider, panel?

**Comment by jefrog1844** [comment]
That sounds fine, I will work on them tomorrow night.

BTW,
Nice pick on the ng lint; bonus since it is integrated with angular already
and if anyone else were to contribute we'd all be on the same linter.  I
tested the tslint plugin for vscode which I’ve been using and it lines up
really well with ng lint.  Now I know why the tslint.json files are in the
project folder. 😃

Thanks for your help,

Jeff

**Comment by jefrog1844** [comment]
I modified tslint to stop modifying/removing the optional parameter on properties when we set a default value.  Now we can set default values that are optional as well.  Also removed the requirement to have the "lib" prefix as we're not using that.  Linting is working great now.

**Comment by jefrog1844** [comment]
Squashed the multiple commits for the 'button' branch

**Pull request merged**